### PR TITLE
Print warnings if alignment names will be changed by IQtree

### PR DESCRIPTION
**Pull request by anna-parker**
### Description of proposed changes
IQtree changes "offending characters" in alignment names to a `'_'` character. Offending characters are defined in: https://github.com/Cibiv/IQ-TREE/blob/6776a95f15a2eccda2aa330497291dc246575995/utils/tools.cpp#L503 (i.e. if the character is not alphanumeric and not in '.', '-' or '_') . Currently, augur will make sure that the offending characters `'/|()*'` are not modified, however all other offending characters will be modified by IQtree. IQtree writes all modifications to a `.log` file. This is not conveyed to the user and alignment information can no longer be used when iterating over the tree. 

I extend augur's current technique for not modifying `'/|()*`' to not modify any offending character. However, treetime might not be able to handle all offending characters. (For example `treetime`'s alignment module reports that `"'"` and `"\'"` are not equivalent - I am not quite sure why -and even if an alignment name with a `'` is not modified the alignment information still cannot be used downstream, for example by `refine`). For offending characters outside of  `'/|()*'` print a warning and a recommendation to change the character to `'_'` to avoid later issues. 

image

### Related issue(s)
https://github.com/nextstrain/augur/issues/1084

### Testing
Modify a sequence name in an alignment file to include an offending character, such as `,@$`, run 
```
augur tree --alignment modified.fasta --tree-builder-args '-ninit 10 -n 4' --output tree_modified.nwk --nthreads 1
```
Check error warning and node names in `tree_modified.nwk`.

### CODE REVIEW

File: augur/tree.py [hunk 1]
```diff
@@ -232,8 +232,22 @@ def random_string(n):
             tmp_line = line
             if line.startswith(">"):
                 num_seqs += 1
-                for c,v in escape_dict.items():
-                    tmp_line = tmp_line.replace(c,v)
+                pos = 1
+                for i in range(1, len(line)):
+                    l = line[i]
+                    #check if l is a character that IQ-tree changes
+                    if l.isalnum()==False and l !='_' and l != '-' and l != '.' and l.isspace()== False:
```
Comment (corneliusroemer) [inline augur/tree.py:239]: You used outdated IQtree code unfortunately, I think there have been recent changes, adding `\` or `/`:
https://github.com/iqtree/iqtree2/blob/3bbc304263cb2f85574a9163e8f2e5c5b597a147/utils/tools.cpp#L585

Maybe link to the IQtree source code in a comment so that in the future, if IQtree changes escaped characters it's easy to find.

You looked up the function in an old repo of IQtree

**Review by corneliusroemer — CHANGES_REQUESTED** [review]

**Comment by corneliusroemer** [comment]
Current IQtree escape characters can be found here: https://github.com/iqtree/iqtree2/blob/3bbc304263cb2f85574a9163e8f2e5c5b597a147/utils/tools.cpp#L585

File: augur/tree.py (hunk 1 above)
Comment (corneliusroemer) [reply augur/tree.py:239]: Actually, the change seems to be that `|` is accepted, but since this seems to be already handled by treetime we don't have an issue here.

But still good to link to the current IQtree source code I think.
```suggestion
                    # See IQtree function renameBool https://github.com/iqtree/iqtree2/blob/3bbc304263cb2f85574a9163e8f2e5c5b597a147/utils/tools.cpp#L585
                    if l.isalnum()==False and l !='_' and l != '-' and l != '.' and l.isspace()== False:
```

Comment (anna-parker) [reply augur/tree.py:240]: Oh great you caught this!

**Comment by tsibley** [comment]
>  For example treetime's alignment module reports that `"'"` and `"\'"` are not equivalent

Can you say more about this?

File: augur/tree.py [hunk 2]
```diff
@@ -232,8 +232,23 @@ def random_string(n):
             tmp_line = line
             if line.startswith(">"):
                 num_seqs += 1
-                for c,v in escape_dict.items():
-                    tmp_line = tmp_line.replace(c,v)
+                pos = 1
+                for i in range(1, len(line)):
+                    l = line[i]
```
Comment (tsibley) [inline augur/tree.py:237]: Python strings are iterable, and since the index is not needed in the loop, this can be simpler:

```suggestion
                for l in line[1:]:
```

File: augur/tree.py [hunk 3]
```diff
@@ -232,8 +232,23 @@ def random_string(n):
...
+                    #check if l is a character that IQ-tree changes
+                    # See IQtree function renameBool https://github.com/iqtree/iqtree2/blob/3bbc304263cb2f85574a9163e8f2e5c5b597a147/utils/tools.cpp#L585
+                    if l.isalnum()==False and l !='_' and l != '-' and l != '.' and l.isspace()== False:
```
Comment (tsibley) [inline augur/tree.py:240]: Hmm. Python's `str.isalnum()` isn't equivalent to the POSIX `isalnum()` used by IQ-TREE. It'd be best to match its behaviour since we're trying to cater to it here.

For `.isspace()`, it seems more sensible to me to avoid modifying anything but the id part (e.g. the part before the first whitespace) of the FASTA defline rather than skipping space mangling. I think IQ-TREE will only care about the id part.

File: augur/tree.py [hunk 4]
```diff
@@ -232,8 +232,23 @@ def random_string(n):
...
+                            print("WARNING: Potentially offending character: \'%s\' detected in taxon name: %s" 
+                                "We recommend replacing offending characters with '_' in the alignment file to avoid issues downstream."
+                                % (l, format(line))
+                            )
+                            escape_dict[l] = f'_{prefix}-{random_string(20)}_'
+                            reverse_escape_dict[escape_dict[l]] = l
+                        tmp_line = tmp_line[:pos] + escape_dict[l] + line[(i+1):]  # replace offending characters
+                        pos = pos + len(escape_dict[l])
+                    else:
+                        pos += 1
```
Comment (tsibley) [inline augur/tree.py:251]: This sort of position tracking and iterative mutation (of `tmp_line`) can get confusing and hard to follow quick, esp. when coming back to it later. I think a functional approach, e.g. with `re.sub()` with a replacement function (or some similar approach), might be more clear. I've sketched out what that looks like: https://gist.github.com/tsibley/8b55c1b721b382cab4c6276442528778

**Comment by tsibley** [comment]
Would also be great to have tests for this behaviour, e.g. in `tests/functional/tree.t` or similar.

File: augur/tree.py (hunk 4 above)
Comment (anna-parker) [reply augur/tree.py:250]: Thank you! I have incorporated your suggestions :-)

**Comment by anna-parker** [comment]
Oh wow its been so long since I worked on this - I will try to add a test and get this ready for review again

**Comment by anna-parker** [comment]
Hi @tsibley I now added some tests - I realized that the replacement of the `'` character still has some unstable behavior - sometimes it gets replaced correctly, other times it gets replaced with `''`. So potentially we should replace these characters with `_` instead (as currently) and just add a warning.

**Comment by genehack** [comment]
@anna-parker it looks like you and @jameshadfield are barking up the same tree -- see https://github.com/nextstrain/augur/pull/1750

**Comment by jameshadfield** [comment]
Hi @anna-parker - apologies for missing this PR when I started working on #1084 recently in #1750.

Both our PRs split the fasta header on whitespace and only use what's before the first whitespace (which is what iq-tree will do anyways) however our escaping approach is quite different.

* This PR escapes any (locale-specific) non-word characters, whereas  #1750 uses a hardcoded list of escape characters.
* This PR prints a warning if single quotes are found whereas #1750 is a fatal error because these are (for me, at least) always escaped when writing out the tree (I think fatal error is appropriate here)
* #1750 also considers backslashes a fatal error as they are often escaped when writing out the tree
* #1750 does a check at the end of `augur tree` to ensure strain names matched

I think #1750 is more complete, but I'm happy to consider `unsafe_chars = re.compile(rb'[^\w.-]', re.LOCALE)` if you think that's the best approach?

**Comment by anna-parker** [comment]
@jameshadfield I also think your approach is more complete - so I am happy to go with that!
